src/app/events/events.service.ts

```diff
@@ -0,0 +1,164 @@
+import { Injectable } from '@angular/core';
+import { Observable, of } from 'rxjs';
+import { map } from 'rxjs/operators';
+import { RequestService } from '@shared/request/request.service';
+import { UtilsService } from '@services/utils.service';
+import { BrowserStorageService } from '@services/storage.service';
+
+/**
+ * @name api
+ * @description list of api endpoint involved in this service
+ * @type {Object}
+ */
+const api = {
+  get: {
+    events: 'api/v2/act/event/list.json',
+    activities: 'api/v2/plan/activity/list.json'
+  },
+  post: {
+    book: 'api/book_events.json'
+  }
+};
+
+export interface Event {
+  id: number;
+  name: string;
+  description: string;
+  location: string;
+  activityId: number;
+  activityName: string;
+  startTime: string;
+  endTime: string;
+  capacity: number;
+  remainingCapacity: number;
+  isBooked: boolean;
+  isPast?: boolean;
+}
+
+export interface Activity {
+  id: number;
+  name: string;
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+
+export class EventsService {
+  constructor(
+    private request: RequestService,
+    private utils: UtilsService,
+    private storage: BrowserStorageService,
+  ) {}
+
+  getEvents(): Observable<any> {
+    return this.request.get(api.get.events, {params: {
+        type: 'activity_session'
+      }})
+      .pipe(map(response => {
+        return this._normaliseEvents(response.data);
+      })
+    );
+  }
+
+  private _normaliseEvents(data): Array<Event> {
+    if (!Array.isArray(data)) {
+      this.request.apiResponseFormatError('Event format error');
+      return [];
+    }
+    let events: Array<Event> = [];
+    data.forEach(event => {
+      if (!this.utils.has(event, 'id') ||
```

Comment:
hi @shawnm0705 , just wondering if this checking for each field is really necessary? 
1) we could add the Event interface for the "event" param above, for eg.
```javascript
data.forEach(event: Event => {
// ...
});
```

2) I think it'll be better if we can just trust API is returning correct API response format, but it's open for discussion.

Reply:
@trtshen I think we can't just put the interface as the type, because API could return more than we need. Since we are doing the data normalisation, I think it's better to do normalisation for every API.
By checking each field, if we change API returns later, we will be able to find out if it still works for App or not. 
By always doing this kind of check and re-assigning data, we can make sure that even we add more return data for API later, it will still work. What do you think?

Reply:
Also if you have noticed, I've changed the variable name from `a_b` to `aB` camelCase, so that we always use camelCase variables

Reply:
Also, this actually gives a list of fields that we are using in the App.
Also if you've noticed that I've converted snake_cased variables to camelCase ones, so that we always use camelCase variables in the code. Sometimes we use different names as well